### DISCUSSION TITLE
How to distribute an asset to partitions?

### DISCUSSION THREAD

**Discussion by psarka** [dagster-io/dagster#12394] [Q&A] [answered]
I need some help with the following computation:

I have an asset X, which is partitioned by a = [1, 2, 3]. 
I would like to produce an asset Y, which would be partitioned by a = [1, 2, 3] and b = ['r', 'g', 'b'].

The computation function is simple: load X[a = k], split into three parts and write to Y[a = k, b = 'r'], Y[a = k, b = 'g'], and Y[a = k, b = 'b']. This should happen in three jobs, one for k = 1, 2, and 3.

I have a solution with 9 jobs - just define X & Y in a straightforward manner, with Y being MultiPartitionsDefinition. But this is not ideal. (In my unsimplified usecase I'm splitting huge files by USA state, and I'd rather not read all of them 50 times).

Is there a way to express this kind of "distribute" pattern? Maybe using jobs and graphs instead of assets?

![image](https://user-images.githubusercontent.com/1212244/219398936-8fe815d2-373b-4bf3-bc3e-af54446baf57.png)

**Comment by sryza** [comment, marked as answer]
I believe that the `SingleDimensionDependencyMapping` that @clairelin135 recently added actually does exactly this. It's not fully exposed yet, but I think we should be able to expose it in next week's release.

@clairelin135 - correct me if I'm wrong here.

**Reply by clairelin135** [reply]
Yes, we should be able to expose a class that does this in next week's release. I have an update out for this that will auto infer the relationship you're looking for without you having to specify this: https://github.com/dagster-io/dagster/pull/12400

**Reply by psarka** [reply]
Amazing! I'll wait for the next week's release!

**Reply by psarka** [reply]
Just to follow up on this, I think the `MultiToSingleDimensionMapping` does not answer my original question, as it still runs one job per one output asset partition, so I get 9 jobs instead of 3.

I am able to implement the behavior I want (1 job that outputs 3 partitions), but I need an auxiliary op. My guess is that current Dagster assumption "asset partitions" == "job partitions" is too fundamental to be changed.

Here is a toy example that distributes one asset (x) to multiple partitions of another asset (y) in one job of the auxiliary op (y_run). That aux op is current implemented as an "asset", but probably could be simplified to an op.

The main thing I don't like about it is that I need to use io manager inside the y_run op. I wish I could just yield a bunch of outputs with their keys and the io manager would "pick them up at the other end".

```python
@asset(
    name="x",
    io_manager_key="touch_and_check",
)
def x_():
    return True


@asset(
    name="y_run",
    io_manager_key="touch_and_check",
    ins={"x": AssetIn(key=["x"])},
)
def y_run_(context: OpExecutionContext, x):
    _ = x

    for partition in ["1", "2", "3"]:
        context.resources.touch_and_check.persist(["y", partition], True)
        yield AssetMaterialization("dim_1__distribute_full_y", partition=partition)

    yield Output(True)


@asset(
    name="y",
    partitions_def=StaticPartitionsDefinition(list("123")),
    io_manager_key="touch_and_check",
    non_argument_deps={"y_run"},
)
def y_():
    pass
```

**Reply by clairelin135** [reply]
Hi @psarka, on Dagit's UI it is possible to execute a run that materializes for multiple partitions: 
<img width="695" alt="image" src="https://user-images.githubusercontent.com/29110579/224791936-711fae08-30d6-417b-b9fb-85d65e2e1eee.png">
which can be enabled by selecting "pass partition ranges in single run". The caveat though is that your IO manager needs to be able to read and write partition ranges. More detailed answer here: https://github.com/dagster-io/dagster/discussions/11653

**Reply by psarka** [reply]
Oh, awesome, I missed that! Will take a look, thanks!

**Reply by psarka** [reply]
@clairelin135 Is there a way to make that checkbox checked for that asset without clicking it on the UI? Otherwise if I have a dag with many assets, and click a global "Materialize all" button, this box remains unchecked and I get many runs.

**Reply by clairelin135** [reply]
Unfortunately not--you'd have to select it on each run that includes that asset
